building-and-packaging/material/packaging_python_slides.md

```diff
@@ -0,0 +1,294 @@
+- pip is itself a [package](https://pypi.org/project/pip/) which is available on PyPI
+- pip is open-source and is developed on [GitHub](https://github.com/pypa/pip)
+
+---
+
+## Using pip
+
+- When `pip install package-name` is run, pip chooses a distribution file and installs it in the environment
+- Doing `python -m pip install package-name` is basically the same as `pip install package-name`
+- pip tracks metadata to allow for easy uninstallation and updating of packages
+- pip is bundled together with Python 3.x, making it even more easier to use
+- pip can install a package from a source distribution (`.tar.gz`) or a wheel distribution (`.whl`)
+- Do not use `sudo pip install`! Various security issues with doing so. Go for `pip install --user <package-name>`
+
+---
+
+## What is PyPI?
+
+- [PyPI](https://pypi.org/) = **The Python Package Index** is a repository of software developed in the Python community
+- PyPI itself is developed on GitHub through another software called [Warehouse](https://github.com/pypa/warehouse)
+- PyPI has a informative [public dashboard](https://p.datadoghq.com/sb/7dc8b3250-85dcf667bd) to show its activity
+- A major advantage is the active maintenance of PyPI and the packages indexed in it
+- Not to be confused with **PyPA** which is Python Packaging Authority, a working group which maintains projects in Python packaging
+
+---
+
+## Example of PyPI package: fenicsprecice
+
+- Having a look at [fenicsprecice](https://pypi.org/project/fenicsprecice/)
+
+---
+
+## Python Enhancement Proposals (PEPs)
+
+- PEP is an evolving design document which provides information regarding new features of Python, new processes and new environments
+- PEPs typically involve concise technical information, which also acts as standardizations
+- Packaging workflows are also standardized through PEPs. Example:
+    - [PEP 427](https://www.python.org/dev/peps/pep-0427/) introduces the built-package format "wheel"
+- It is always a good idea to look for **PEP** conventions before trying out processes
+
+---
+
+## File structure for code packaging 1/2
+
+- Most essential component of packaging is file structure. Basic file structure is as follows:
+
+```bash
+generic_folder_name/
+└── src/
+    └── package_name/
+        ├── __init__.py
+        └── solver.py
+```
+
+- The file `__init__.py` is required to import the `package_name/` as a package. This file is mostly empty
+- `solver.py` will contain the code developed by the user
+
+---
+
+## File structure for code packaging 2/2
+
+- Once the file structuring is complete, the repository will look like:
+
+```bash
+generic_folder_name/
+├── LICENSE
+├── setup.py
+├── README.md
+├── src/
+│   └── package_name/
+│       ├── __init__.py
+│       └── solver.py
+└── tests/
+```
+
+<span>
+Lets have a detailed look at individual files in this folder structure
+<!-- .element: class="fragment" data-fragment-index="1" --></span>
+
+---
+
+## Additional options in setup.py 1/2
```

Comment:
I am a bit confused why this additional information does not come earlier when `setup.py` is introduced. Can I only use these additional options if I use pip or what is the reason for that? Otherwise it might make sense to move this to the beginning of the lecture to keep the part about `setup.py` well connected. Then you could refer here to the previously shown information about the `setup.py` without repeating it.